respy/method_of_simulated_moments.py

```diff
@@ -0,0 +1,361 @@
+        empirical_moments. Used to weight squared moment errors.
+    return_scalar : bool
+        Indicates whether to return moment error vector (False) or weighted square
+        product of moment error vector (True).
+
+    Returns
+    -------
+    out : pandas.Series or float
+        Scalar or moment error vector depending on value of return_scalar.
+
+    """
+    empirical_moments = copy.deepcopy(empirical_moments)
+
+    df_sim = simulate(params)
+    simulated_moments = _calculate_moments(df_sim, calc_moments)
+
+    simulated_moments = _reindex_simulated_moments(empirical_moments, simulated_moments)
+
+    simulated_moments = _replace_missing_values(simulated_moments, replace_nans)
+
+    flat_empirical_moments = _flatten_index(empirical_moments)
+    flat_simulated_moments = _flatten_index(simulated_moments)
+
+    moment_errors = flat_empirical_moments - flat_simulated_moments
+
+    # Return moment errors as indexed DataFrame or calculate weighted square product of
+    # moment errors depending on return_scalar.
+    if return_scalar is False:
+        out = moment_errors
+    else:
+        out = moment_errors.T @ weighting_matrix @ moment_errors
+
+    return out
+
+
+def get_diag_weighting_matrix(empirical_moments, weights=None):
+    """Create a diagonal weighting matrix from weights.
+
+    Parameters
+    ----------
+    empirical_moments : pandas.DataFrame or pandas.Series or dict or list
+        Contains the empirical moments calculated for the observed data. Moments should
+        be saved to pandas.DataFrame or pandas.Series that can either be passed to the
+        function directly or as items of a list or dictionary.
+    weights : pandas.DataFrame or pandas.Series or dict or list
+        Contains weights (usually variances) of empirical moments. Must match structure
+        of empirical_moments i.e. if empirical_moments is a list of pandas.DataFrames,
+        weights be list of pandas.DataFrames as well where each DataFrame entry contains
+        the weight for the corresponding moment in empirical_moments.
+
+    Returns
+    -------
+    numpy.ndarray
+        Array contains a diagonal weighting matrix.
+
+    """
+    empirical_moments = copy.deepcopy(empirical_moments)
+    empirical_moments = _harmonize_input(empirical_moments)
+
+    # Use identity matrix if no weights are specified.
+    if weights is None:
+        weights = copy.deepcopy(empirical_moments)
+        for df in weights:
+            df[:] = 1
+
+    # Harmonize input weights.
+    else:
+        weights = copy.deepcopy(weights)
+        weights = _harmonize_input(weights)
+
+    flat_weights = _flatten_index(weights)
+    flat_empirical_moments = _flatten_index(empirical_moments)
+
+    # Sort index in weights according to empirical moments.
+    flat_weights = flat_empirical_moments.join(flat_weights, lsuffix="_emp").iloc[:, 1:]
+
+    # Return diagonal weighting matrix.
+    return np.diag(flat_weights.iloc[:, 0])
```

Comment:
I am not sure what this ``iloc`` should do? At this point, ``flat_weights`` should be a series and if you want a Numpy array use ``.to_numpy()``, but I think ``np.diag`` already return a two-dim array from a series.